chain/chain/src/chain.rs

```diff
@@ -3890,54 +3929,71 @@ impl Chain {
         let _span = tracing::debug_span!(target: "chain", "apply_chunks_preprocessing").entered();
         let prev_chunk_headers =
             Chain::get_prev_chunk_headers(self.epoch_manager.as_ref(), prev_block)?;
-        block
-            .chunks()
-            .iter()
-            .zip(prev_chunk_headers.iter())
-            .enumerate()
-            .filter_map(|(shard_id, (chunk_header, prev_chunk_header))| {
-                // XXX: This is a bit questionable -- sandbox state patching works
-                // only for a single shard. This so far has been enough.
-                let state_patch = state_patch.take();
-                let update_shard_job = self.get_update_shard_job(
+
+        let mut jobs = vec![];
+        for (shard_id, (chunk_header, prev_chunk_header)) in
+            block.chunks().iter().zip(prev_chunk_headers.iter()).enumerate()
+        {
+            // XXX: This is a bit questionable -- sandbox state patching works
+            // only for a single shard. This so far has been enough.
+            let state_patch = state_patch.take();
+
+            // Insert job into list of all jobs to run, if job is valid.
+            let mut insert_job = |job: Result<Option<UpdateShardJob>, Error>| {
```

Comment:
nit: this looks weird 🙃  can we just accumulate `job_results: Vec<Result<Option<UpdateShardJob>, Error>>` and then extract jobs and invalid_chunks after this loop?